util/src/io/mod.rs

```diff
@@ -0,0 +1,1089 @@
+//!   and does not wrap another error. This is because we need to be able to
+//!   construct I/O errors even when we cannot allocate.
+mod cursor;
+mod error;
+mod impls;
+mod initializer;
+pub mod prelude;
+mod util;
+pub use self::cursor::Cursor;
+pub use self::error::{Error, ErrorKind, Result};
+pub use self::initializer::Initializer;
+pub use self::util::{copy, empty, repeat, sink, Empty, Repeat, Sink};
+use core::{cmp, fmt, slice, str};
+
+#[cfg(feature = "alloc")]
+use alloc::{string::String, vec::Vec};
+// FIXME(eliza): calculate based on arch?
+const DEFAULT_BUF_SIZE: usize = 256;
+
+pub trait Read {
+    /// Pull some bytes from this source into the specified buffer, returning
+    /// how many bytes were read.
+    ///
+    /// This function does not provide any guarantees about whether it blocks
+    /// waiting for data, but if an object needs to block for a read but cannot
+    /// it will typically signal this via an [`Err`] return value.
+    ///
+    /// If the return value of this method is [`Ok(n)`], then it must be
+    /// guaranteed that `0 <= n <= buf.len()`. A nonzero `n` value indicates
+    /// that the buffer `buf` has been filled in with `n` bytes of data from this
+    /// source. If `n` is `0`, then it can indicate one of two scenarios:
+    ///
+    /// 1. This reader has reached its "end of file" and will likely no longer
+    ///    be able to produce bytes. Note that this does not mean that the
+    ///    reader will *always* no longer be able to produce bytes.
+    /// 2. The buffer specified was 0 bytes in length.
+    ///
+    /// No guarantees are provided about the contents of `buf` when this
+    /// function is called, implementations cannot rely on any property of the
+    /// contents of `buf` being true. It is recommended that *implementations*
+    /// only write data to `buf` instead of reading its contents.
+    ///
+    /// Correspondingly, however, *callers* of this method may not assume any guarantees
+    /// about how the implementation uses `buf`. The trait is safe to implement,
+    /// so it is possible that the code that's supposed to write to the buffer might also read
+    /// from it. It is your responsibility to make sure that `buf` is initialized
+    /// before calling `read`. Calling `read` with an uninitialized `buf` (of the kind one
+    /// obtains via [`MaybeUninit<T>`]) is not safe, and can lead to undefined behavior.
+    ///
+    /// [`MaybeUninit<T>`]: ../mem/union.MaybeUninit.html
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters any form of I/O or other error, an error
+    /// variant will be returned. If an error is returned then it must be
+    /// guaranteed that no bytes were read.
+    ///
+    /// An error of the [`ErrorKind::Interrupted`] kind is non-fatal and the read
+    /// operation should be retried if there is nothing else to do.
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
```

Comment:
Given that we don't need to maintain compatibility with `std`, it would be possible to make `read` take a signature more like:

```rust
fn read<'a>(&mut self, buf: &'a mut [MaybeUninit<u8>]) -> Result<(&'a mut [u8], &'a mut [MaybeUninit<u8>])>;
```

Which would explicitly encode the initialization range.
That being said, that signature is _way_ noisier, and the current behaviour in rust has to stay sound for other reasons, so we can probably get away with copying with `std` does.

Reply:
IMHO `[MaybeUninit<u8>]` is a better approach, but duplicating std exactly lets us copy a lot more code from std without modifying it...I'm open to changing this in a follow-up.